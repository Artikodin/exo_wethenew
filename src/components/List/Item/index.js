import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

import "./style.css";

/**
 * Item
 *
 * This components is in charge to display an item of the list
 *
 * @param {number} id - the product id
 * @param {string} source - the product main image source
 * @param {string} title - the product title
 * @param {string} price - the product price
 *
 * @return {jsx}
 */

const Item = ({ id, source, title, price }) => {
  return (
    <li className="card" data-testid="list-item">
      <Link to={`/product/${id}`} className="custom-link">
        <img className="card__image" src={source} alt={title} />
        <h3 className="card__title">{title}</h3>
        <p className="card__price">{price} $</p>
      </Link>
    </li>
  );
};

const { string } = PropTypes;
Item.propTypes = { source: string, title: string, price: string };
Item.defaultProps = { source: "", title: "", price: "" };

export default Item;
