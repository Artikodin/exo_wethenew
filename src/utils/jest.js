import React from "react";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router, Route } from "react-router-dom";

/**
 * renderWithRouter
 *
 * Render a component with a mocked router. 
 * Usefull when you want to test a component containing react router component.
 * 
 * @param {node} children - react component you want to add routing logic
 * @param {object} routeParams - different parameter you use to simulate a routing behaviour
 *
 * @return {node} A react router mock
 */

export const renderWithRouter = (
  children,
  {
    url = "/",
    route: { path = "/", isExact = false } = {},
    history = createMemoryHistory({ initialEntries: [url] }),
  } = {}
) => {
  return {
    ...render(
      <Router history={history}>
        <Route path={path} exact={isExact}>
          {children}
        </Route>
      </Router>
    ),
    history,
  };
};
