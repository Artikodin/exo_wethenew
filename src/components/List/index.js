import React from "react";
import PropTypes from "prop-types";

import Item from "./Item";
import "./style.css";

/**
 * List
 *
 * This components is in charge to display a list of product
 *
 * @param {array} products - array of products
 *
 * @return {jsx}
 */

const List = ({ products }) => {
  return (
    <ul className="list">
      {products.map(({ id, name, price, shoes_pictures }) => (
        <Item
          key={id}
          id={id}
          price={price}
          title={name}
          source={shoes_pictures[0]}
        />
      ))}
    </ul>
  );
};

const { array } = PropTypes;
List.propTypes = { products: array };
List.defaultProps = { products: [] };

export default List;
