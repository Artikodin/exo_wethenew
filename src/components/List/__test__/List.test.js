import React from "react";
import { cleanup } from "@testing-library/react";
import { renderWithRouter } from "../../../utils/jest";

import List from "..";

const mockProducts = [
  {
    name: "Air Jordan Mid Iridescent ReflectiveWhite",
    brand: "Jordan",
    price: "230",
    size: ["36", "37", "41"],
    shoes_pictures: [
      "https://cdn.shopify.com/s/files/1/2358/2817/products/Wethenew-Sneakers-France-Air-jordan-1-Mid-Iridescent-Reflective-White-CK6587-100-1_2000x.png?v=1583933418",
      "https://cdn.shopify.com/s/files/1/2358/2817/products/Wethenew-Sneakers-France-Air-jordan-1-Mid-Iridescent-Reflective-White-CK6587-100-2_2000x.png?v=1583933426",
    ],
    Description:
      "The Air Jordan 1 Mid collection is growing! The brand in the Jumpman offers us a new declination intended for women, in a colour that at first glance seems rather sober but which is much more worked once its reflective elements are visible.",
  },
  {
    name: "Air Force 1 '07 Barely Grape",
    brand: "Nike",
    price: "170",
    size: ["36", "37"],
    shoes_pictures: [
      "https://cdn.shopify.com/s/files/1/2358/2817/products/Wethenew-Sneakers-France-Nike-Air-Force-1-_07-Barely-Grape-CU3449-100-1_2000x.png?v=1583921569",
      "https://cdn.shopify.com/s/files/1/2358/2817/products/Wethenew-Sneakers-France-Nike-Air-Force-1-_07-Barely-Grape-CU3449-100-1_2000x.png?v=1583921569",
    ],
    Description:
      "Nike's most iconic silhouette makes a comeback in a pale colourMauve, ideal for the upcoming summer season. Designed by Bruce Kilgore in 1982, this is the first pair with Nike AIR technology, which was originally intended for basketball.",
  },
  {
    name: "Air Max97 Black",
    brand: "Nike",
    price: "200",
    size: ["36", "37", "41", "45", "49"],
    shoes_pictures: [
      "https://cdn.shopify.com/s/files/1/2358/2817/products/Wethenew-Sneakers-France-Nike-Air-Max-97-Black-921733-006-2_2000x.png?v=1583939000",
      "https://cdn.shopify.com/s/files/1/2358/2817/products/Wethenew-Sneakers-France-Nike-Air-Max-97-Black-921733-006-3_2000x.png?v=1583939007",
    ],
    Description:
      "Nike unveils a new iteration of one of its flagship models: the Nike Air Max 97A pair with multiple inspirations, especially water represented by the characteristic lines of the model.",
  },
];

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

describe("<List />", () => {
  test("should render withouth a problem", () => {
    const { container } = renderWithRouter(<List />);

    expect(container.firstChild).toMatchSnapshot();
  });
  
  test("should list exact number of item", () => {
    const { queryAllByTestId } = renderWithRouter(
      <List products={mockProducts} />
      );
      
      const listItems = queryAllByTestId("list-item");
      expect(listItems.length).toBe(3);
  });
});
