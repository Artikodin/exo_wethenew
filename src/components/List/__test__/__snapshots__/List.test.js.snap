// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should list exact number of item 1`] = `3`;

exports[`<List /> should render withouth a problem 1`] = `
<ul
  class="list"
/>
`;

exports[`<PageLayout /> should render withouth a problem 1`] = `
<ul
  class="list"
/>
`;
